apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableMutateRowsRequestManager.cs

```diff
@@ -0,0 +1,212 @@
+﻿// Copyright 2017 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+using System.Linq;
+using Google.Api.Gax.Grpc;
+using Grpc.Core;
+
+namespace Google.Cloud.Bigtable.V2
+{
+    /// <summary>
+    /// Performs retries for <see cref="BigtableClient.MutateRows(MutateRowsRequest, CallSettings)"/> operations.
+    /// </summary>
+    internal class BigtableMutateRowsRequestManager
+    {
+        private readonly MutateRowsRequest _originalRequest;
+
+        /// <summary>
+        /// The new request to send. This starts as the original request. If retries occur, this
+        /// request will contain the subset of Entries that need to be retried.
+        /// </summary>
+        internal MutateRowsRequest RetryRequest { get; private set; }
+        
+        /// <summary>
+        /// When doing retries, the retry sends a partial set of the original Entries that failed with a
+        /// retryable status. This array contains a mapping of indices from the <see cref="RetryRequest"/> 
+        /// to <see cref="_originalRequest"/>.
+        /// </summary>
+        private int[] _mapToOriginalIndex;
+
+        /// <summary>
+        /// This list tracks failed RPC status codes thats should be retried
+        /// </summary> 
+        private readonly HashSet<int> _rertriableCodes;
+
+        /// <summary>
+        /// This array tracks the cumulative set of results across all RPC requests.
+        /// Maps to the original request's entries indices.
+        /// </summary> 
+        private readonly Rpc.Status[] _results;
+
+        private static Rpc.Status StatusInternal => new Rpc.Status
+        {
+            Code = (int)StatusCode.Internal,
+            Message = "Response was not returned for this index"
+        };
+
+        private bool _messageIsInvalid;
+
+        private static MutateRowsResponse.Types.Entry CreateEntry(int i, Rpc.Status status)
+        {
+            return new MutateRowsResponse.Types.Entry { Index = i, Status = status };
+        }
+
+        /// <summary>
+        /// Contains possible processing statuses
+        /// </summary>
+        internal enum ProcessingStatus
```

Comment:
(repeating your comment here from #1797)
_Based on the ProcessingStatus the "Retryer" utility will decide its next step, for example,_

* _SUCCESS => BuildResponse() => return full response to user_
* _RETRYABLE => RetryRequest and retry or (if exhausted all retry attempts) BuildResponse() => return full response to user._
* _NOT_RETRYABLE =>BuildResponse() => return full response to user._
* _INVALID => return message to user that something went wrong_

So like I was saying, it seems like `SUCCESS` is unnecessary. For the purposes of this class and callers to it, `SUCCESS` and `NOT_RETRYABLE` appear to be semantically equivalent. They both mean "BuildResponse() => return full response to user". All successful mutations is just one case of something that shouldn't be retried.